File: python/src/threads/agents/stop.py
"""An early end stops the run's children (spec/schema/README.md, "Run completion"): a run that
ended failed, budget_exhausted or handed_off gives each background child it waits on (its own,
and an earlier run's it relaunched) a durable tree cancel, waits the step each needs to stop, and
records their ends with no wake. A cancel of the lead during that wait ends it."""

import asyncio
from collections.abc import Sequence
from typing import Final

from threads.agents.background import Background
from threads.agents.scope import Scope
from threads.log import AgentSpawnedEvent, CancelRequestedEvent, Event, UserInputEvent
from threads.loop.drive import drive
from threads.loop.runtime import Halt, Runtime
from threads.reduce.run_end import RunEnd, RunStatus, ended_otherwise, run_end
from threads.thread.tree import bar_child

REASON: Final = "parent run ended"
RETRY_S: Final = 0.05
"""How often a child not barred yet is tried again: its thread may not exist at the first try."""


def _latest(events: Sequence[Event]) -> RunEnd | None:
    request = next((e for e in reversed(events) if isinstance(e, UserInputEvent)), None)
    return None if request is None else run_end(events, request.event_id)


def run_status(events: Sequence[Event]) -> RunStatus:
    """The status of the latest request's run."""
    end = _latest(events)
    return "running" if end is None else end.status


def _waited_on(rt: Runtime, bg: Background) -> list[AgentSpawnedEvent]:
    """The background children this run waits on and that have not reported: its own, and any
    earlier run's it relaunched."""
    request = next((e for e in reversed(rt.events) if isinstance(e, UserInputEvent)), None)
    fold = rt.fold

    def waited(e: AgentSpawnedEvent) -> bool:
        child = e.data.child_thread_id
        if fold.children.get(child) is not False:
            return False
        run = fold.team.spawns.get(e.data.call_id)
        root = None if run is None else run.root
        mine = request is not None and root is not None and root[1] == request.event_id
        return mine or bg.has(child)

    return [
        e
        for e in rt.events
        if isinstance(e, AgentSpawnedEvent) and e.data.mode == "background" and waited(e)
    ]


def _cancelled_after(events: Sequence[Event], seq: int) -> bool:
    """A thread or tree cancel of the lead after `seq`."""
    return any(
        isinstance(e, CancelRequestedEvent) and e.seq > seq and e.data.scope != "turn"
        for e in events
    )


async def stop_children[D](scope: Scope[D], rt: Runtime, bg: Background, halt: Halt) -> Halt:
    # A cancel of the lead after the run's deciding event ends the wait.
    end = _latest(rt.events)
    decided = rt.fold.seq if end is None or end.at is None else rt.events[end.at].seq
    barred: set[str] = set()
    left = _waited_on(rt, bg)
    while left and ended_otherwise(run_status(rt.events)):
        # Taken before the checks: a cancel that lands after them still wakes the wait.
        moved = rt.writer.moved()
        if _cancelled_after(rt.events, decided):
            moved.close()
            return halt
        for spawned in left:
            child = spawned.data.child_thread_id
            if child not in barred and await bar_child(scope.store, child, scope.principal, REASON):
                barred.add(child)
        running = [s for s in left if s.data.child_thread_id in bg.running]
        if not running and not bg.ended:
            moved.close()
            return halt
        if bg.ended:
            moved.close()
        else:
            unbarred = any(s.data.child_thread_id not in barred for s in left)
            # ponytail: re-sends every 50 ms until each child has its barrier; a start
            # notification would do it without a timer.
            timer = (asyncio.sleep(RETRY_S),) if unbarred else ()
            await bg.next_end(moved, *timer)
        halt = await drive(rt)
        left = _waited_on(rt, bg)
    return halt

File: typescript/packages/core/src/loop/agents/stop.ts
import { endedOtherwise, type RunStatus, runEnd } from "../../reduce/run-end";
import type { LoopEnd } from "../run";
import { runLoop } from "../run";
import type { Session } from "../session";
import type { Spawned } from "./spawn";

// An early end stops the run's children (spec/schema/README.md, "Run completion"): a run that
// ended failed, budget_exhausted or handed_off gives each of its background children that has
// not reported a durable tree cancel, waits the step each needs to stop, and records their ends
// with no wake, so no running child or pending_wakes row outlives run().

const REASON = "parent run ended";
/** How often an unstarted child is barred again: its thread may not exist at the first try. */
const RETRY_MS = 50;

/** The status of the latest request's run. */
export function runStatus(s: Session): RunStatus {
  const request = s.events.findLast((e) => e.type === "user_input");
  return request === undefined
    ? "running"
    : runEnd(s.events, request.event_id).status;
}

/** This run's background children still running or not yet recorded. */
function unreported(s: Session): readonly Spawned[] {
  const request = s.events.findLast((e) => e.type === "user_input");
  return s.events.filter(
    (e): e is Spawned =>
      e.type === "agent_spawned" &&
      e.data.mode === "background" &&
      s.fold.children.get(e.data.child_thread_id) === "running" &&
      s.fold.team.spawns.get(e.data.call_id)?.root?.event === request?.event_id,
  );
}

export async function stopChildren(
  s: Session,
  first: LoopEnd,
): Promise<LoopEnd> {
  let end = first;
  for (let left = unreported(s); left.length > 0; left = unreported(s)) {
    if (end.kind === "halted" || !endedOtherwise(runStatus(s))) return end;
    for (const spawned of left)
      await s.config.agents
        ?.subagent(spawned.data.agent_name)
        ?.stop(spawned.data.child_thread_id, s.config.principal, REASON);
    const running = left.flatMap((e) => s.background.get(e.data.call_id) ?? []);
    // A child launched but not yet started gets its barrier on a later pass.
    // ponytail: polls every 50 ms; a start notification would do it without a timer.
    const tick = Promise.withResolvers<void>();
    const timer = setTimeout(tick.resolve, RETRY_MS);
    if (s.finished.size === 0 && running.length > 0)
      await Promise.race([...running, s.moved(), tick.promise]);
    clearTimeout(timer);
    if (running.length === 0 && s.finished.size === 0) return end;
    end = await runLoop(s);
  }
  return end;
}
